fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java

```diff
@@ -907,63 +919,81 @@ public void rollEditLog() {
-            throw new Error("Fatal Error : no editLog stream");
-        }
-
-        long start = System.currentTimeMillis();
-
-        Preconditions.checkState(GlobalStateMgr.getCurrentState().isMaster(),
-                "non-master fe can not write bdb log");
-
-        try {
-            journal.write(op, writable);
-        } catch (Exception e) {
-            LOG.error("Fatal Error : write stream Exception", e);
-            System.exit(-1);
-        }
+    protected void logEdit(short op, Writable writable) {
+        logEdit(op, writable, -1);
+    }
 
-        // get a new transactionId
-        txId++;
 
-        // update statistics
-        long end = System.currentTimeMillis();
-        numTransactions++;
-        totalTimeTransactions += (end - start);
+    protected void logEdit(short op, Writable writable, long maxWaitIntervalMs) {
+        long start = System.currentTimeMillis();
+        JournalQueueEntity entity = putLogToQueue(op, writable, maxWaitIntervalMs);
+        waitLogConsumed(entity);
         if (MetricRepo.isInit) {
-            MetricRepo.HISTO_EDIT_LOG_WRITE_LATENCY.update((end - start));
+            MetricRepo.HISTO_EDIT_LOG_WRITE_LATENCY.update((System.currentTimeMillis() - start));
         }
+    }
 
-        if (LOG.isDebugEnabled()) {
-            LOG.debug("nextId = {}, numTransactions = {}, totalTimeTransactions = {}, op = {}",
-                    txId, numTransactions, totalTimeTransactions, op);
-        }
+    /**
+     * put log in queue and return immediately
+     */
+    protected JournalQueueEntity putLogToQueue(short op, Writable writable, long maxWaitIntervalMs) {
+        DataOutputBuffer buffer = new DataOutputBuffer(OUTPUT_BUFFER_INIT_SIZE);
 
-        if (txId >= Config.edit_log_roll_num) {
-            LOG.info("txId {} is equal to or larger than edit_log_roll_num {}, will roll edit.",
-                    txId, Config.edit_log_roll_num);
-            rollEditLog();
-            txId = 0;
+        // 1. serialized
+        try {
+            JournalEntity entity = new JournalEntity();
+            entity.setOpCode(op);
+            entity.setData(writable);
+            entity.write(buffer);
+        } catch (IOException e) {
+            // The old implement swallow exception like this
+            LOG.info("failed to serialized: {}", e);
         }
-
-        if (MetricRepo.isInit) {
-            MetricRepo.COUNTER_EDIT_LOG_WRITE.increase(1L);
+        JournalQueueEntity entity = new JournalQueueEntity(op, buffer, maxWaitIntervalMs);
+
+        /*
+         * for historical reasons, logEdit is not allowed to raise Exception, which is really unreasonable to me.
+         * This PR will continue to swallow exception and retry till the end of the world like before.
+         * Hope some day we'll fix it.
+         */
+        // 2. put to queue
+        int cnt = 0;
+        while (true) {
+            try {
+                if (cnt != 0) {
+                    Thread.sleep(1000);
```

Comment:
Why so many sleep? Why not a condition variable?

Reply:
1. We only go to this branch when InterruptedException is raised, for example, if this thread is interrupt as `testNoWaitInterrupt()` shows. In this scenerio, a condition variable will also raise a InterruptedException.
2. I actually think that this exception should be raised so that the caller can know it's time to clean up and quit the context. But as I commented above, this function is not allowed to raise Exception for historical reasons. So we can only wait and loop forever as the old implematation did.